admin/server/reports.go

```diff
@@ -57,48 +61,38 @@ func (s *Server) GetReportMeta(ctx context.Context, req *adminv1.GetReportMetaRe
-	if len(req.EmailRecipients) == 0 {
-		return &adminv1.GetReportMetaResponse{
-			BaseUrls: &adminv1.GetReportMetaResponse_URLs{
-				OpenUrl:   s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportOpen(org.Name, proj.Name, req.Report, req.ExecutionTime.AsTime()),
-				ExportUrl: s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportExport(org.Name, proj.Name, req.Report, ""),
-				EditUrl:   s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportEdit(org.Name, proj.Name, req.Report, ""),
-			},
-		}, nil
-	}
+	urls := make(map[string]*adminv1.GetReportMetaResponse_URLs)
 
-	var externalEmails []string
-	for _, email := range req.EmailRecipients {
-		_, err := s.admin.DB.FindUserByEmail(ctx, email)
-		if err != nil {
-			if errors.Is(err, database.ErrNotFound) {
-				externalEmails = append(externalEmails, email)
-			} else {
-				return nil, fmt.Errorf("failed to find user by email: %w", err)
-			}
-		}
+	var recipients []string
+	recipients = append(recipients, req.EmailRecipients...)
+	if req.AnonRecipients {
+		// add empty email for slack and other notifiers token
+		recipients = append(recipients, "")
 	}
 
-	externalUrls := make(map[string]*adminv1.GetReportMetaResponse_URLs, len(externalEmails))
-	emailTokens, err := s.reconcileMagicTokensForExternalEmails(ctx, proj.OrganizationID, proj.ID, req.Report, req.OwnerId, externalEmails)
+	tokens, err := s.createMagicTokens(ctx, proj.OrganizationID, proj.ID, req.Report, req.OwnerId, req.MetricsView, req.Explore, req.Canvas, recipients)
 	if err != nil {
 		return nil, fmt.Errorf("failed to issue magic auth tokens: %w", err)
 	}
 
-	for email, token := range emailTokens {
-		externalUrls[email] = &adminv1.GetReportMetaResponse_URLs{
-			ExportUrl: s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportExport(org.Name, proj.Name, req.Report, token),
-			EditUrl:   s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportEdit(org.Name, proj.Name, req.Report, token),
+	for _, recipient := range req.EmailRecipients {
+		urls[recipient] = &adminv1.GetReportMetaResponse_URLs{
+			OpenUrl:        s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportOpen(org.Name, proj.Name, req.Report, tokens[recipient], req.ExecutionTime.AsTime()),
+			ExportUrl:      s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportExport(org.Name, proj.Name, req.Report, tokens[recipient]),
+			UnsubscribeUrl: s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportUnsubscribe(org.Name, proj.Name, req.Report, tokens[recipient], recipient),
+		}
+	}
+
+	if req.AnonRecipients {
+		urls[""] = &adminv1.GetReportMetaResponse_URLs{
+			OpenUrl:        s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportOpen(org.Name, proj.Name, req.Report, tokens[""], req.ExecutionTime.AsTime()),
+			ExportUrl:      s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportExport(org.Name, proj.Name, req.Report, tokens[""]),
+			UnsubscribeUrl: s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportUnsubscribe(org.Name, proj.Name, req.Report, tokens[""], ""),
```

Comment:
How is this different from the previous implementation that looped over `tokens`? Since `""` is in `tokens`, could it automatically be added in the same loop?

Reply:
sure